oauth2/hook.go

```diff
@@ -163,3 +164,393 @@ func RefreshTokenHook(reg interface {
 		return nil
 	}
 }
+
+// AuthorizationCodeHookRequest is the request body sent to the authorization code token hook.
+//
+// swagger:model authorizationCodeHookRequest
+type AuthorizationCodeHookRequest struct {
+	// Subject is the identifier of the authenticated end-user.
+	Subject string `json:"subject"`
+	// Session is the request's session..
+	Session *Session `json:"session"`
+	// Requester is a token endpoint's request context.
+	Requester Requester `json:"requester"`
+	// ClientID is the identifier of the OAuth 2.0 client.
+	ClientID string `json:"client_id"`
+	// GrantedScopes is the list of scopes granted to the OAuth 2.0 client.
+	GrantedScopes []string `json:"granted_scopes"`
+	// GrantedAudience is the list of audiences granted to the OAuth 2.0 client.
+	GrantedAudience []string `json:"granted_audience"`
+}
+
+// AuthorizationCodeHookResponse is the response body received from the authorization code token hook.
+//
+// swagger:model authorizationCodeHookResponse
+type AuthorizationCodeHookResponse struct {
+	// Session is the session data returned by the hook.
+	Session consent.AcceptOAuth2ConsentRequestSession `json:"session"`
+}
+
+// AuthorizationCodeHook is an AccessRequestHook called for `authorization_code` grant type.
+func AuthorizationCodeHook(reg interface {
```

Comment:
I am unsure about how this kind os factorisation should be done.
My first instinct is to abuse generics. Is that OK with you ?

Reply:
@fehrnah Every function here has 2 points that are different - get a `hookURL` and check the grant type from the requester. The rest of the function body is the same for all new functions. Basically lines 213 - 299 could be moved into a separate function and reused.

Reply:
With that approach you could also get rid of repeating struct definitions around `*Request` and `*Response` since they are the same (at least for now)

Reply:
Thank you for the input.
This has not been my priority for the las few moth, but I'm still interested in getting things through.

I made the changes you suggested and things are much better !

Reply:
@fehrnah Awesome work! Could please add a hook for https://www.ory.sh/docs/hydra/guides/jwt#exchanging-jwts-for-access-tokens?

Reply:
> @fehrnah Awesome work! Could please add a hook for https://www.ory.sh/docs/hydra/guides/jwt#exchanging-jwts-for-access-tokens?

I think this is what I called JWTProfile.
I renamed it and related conf to JWTBearer to be consistant.

So after this, tthere are hooks for the following grant_types:
- refresh_token
- authorization_code
- client_credentials
- urn:ietf:params:oauth:grant-type:jwt-bearer

Are there more I should support ?

Reply:
No, I think it is fine to support the ones you added and skip the implicit flow. @aeneasr what do you think?